integration/distinct_compat_test.go

```diff
@@ -0,0 +1,157 @@
+			t.Helper()
+
+			if tc.skipForTigris != "" {
+				setup.SkipForTigrisWithReason(t, tc.skipForTigris)
+			}
+
+			t.Parallel()
+
+			filter := tc.filter
+			require.NotNil(t, filter, "filter should be set")
+
+			var nonEmptyResults bool
+			for i := range targetCollections {
+				targetCollection := targetCollections[i]
+				compatCollection := compatCollections[i]
+				t.Run(targetCollection.Name(), func(t *testing.T) {
+					t.Helper()
+
+					targetRes, targetErr := targetCollection.Distinct(ctx, tc.field, tc.filter)
+					compatRes, compatErr := compatCollection.Distinct(ctx, tc.field, tc.filter)
+
+					if targetErr != nil {
+						t.Logf("Target error: %v", targetErr)
+						targetErr = UnsetRaw(t, targetErr)
+						compatErr = UnsetRaw(t, compatErr)
+						assert.Equal(t, compatErr, targetErr)
+						return
+					}
+					require.NoError(t, compatErr, "compat error; target returned no error")
+
+					t.Logf("Compat (expected) result: %v", compatRes)
+					t.Logf("Target (actual)   result: %v", targetRes)
+
+					require.Equal(t, len(compatRes), len(targetRes))
+
+					// We can't check the exact data types because they might be different.
+					// For example, if targetRes is [float64(0), int32(1)] and compatRes is [int64(0), int64(1)],
+					// we consider them equal. If different documents use different types to store the same value
+					// in the same field, it's hard to predict what type will be returned by distinct.
+					// This is why we iterate through results and use assert.EqualValues instead of assert.Equal.
+					for i := range compatRes {
+						assert.EqualValues(t, compatRes[i], targetRes[i])
+					}
+
+					if targetRes != nil || compatRes != nil {
+						nonEmptyResults = true
+					}
+				})
+			}
+
+			switch tc.resultType {
+			case nonEmptyResult:
+				assert.True(t, nonEmptyResults, "expected non-empty results")
+			case emptyResult:
+				assert.False(t, nonEmptyResults, "expected empty results")
+			default:
+				t.Fatalf("unknown result type %v", tc.resultType)
+			}
+		})
+	}
+}
+
+func TestDistinctCompat(t *testing.T) {
+	t.Parallel()
+
+	testCases := map[string]distinctCompatTestCase{
+		"EmptyField": {
+			field:      "",
+			filter:     bson.D{},
+			resultType: emptyResult,
+		},
+		"IDAny": {
+			field:  "_id",
+			filter: bson.D{},
+		},
+		"IDString": {
+			field:  "_id",
+			filter: bson.D{{"_id", "string"}},
+		},
+		"IDNotExists": {
+			field:  "_id",
+			filter: bson.D{{"_id", "count-id-not-exists"}},
+		},
+		"VArray": {
+			field:         "v",
+			filter:        bson.D{{"v", bson.D{{"$type", "array"}}}},
+			skipForTigris: "https://github.com/FerretDB/FerretDB/issues/1704",
+		},
+		"VAny": {
+			field:  "v",
+			filter: bson.D{},
+		},
+		"NonExistentField": {
+			field:  "field-not-exists",
+			filter: bson.D{},
+		},
+		"DotNotation": {
+			field:  "v.foo",
+			filter: bson.D{},
+		},
+		"DotNotationArray": {
+			field:  "v.array.0",
+			filter: bson.D{},
+		},
```

Comment:
I suggest one more edge test case. This one has challenge in other tests and was skipped https://github.com/FerretDB/FerretDB/blob/main/integration/update_field_compat_test.go#L208

```
field: "v.0.foo",
filter: bson.D{},
resultType: emptyResult,
```

Reply:
I added a test case, but here it's not `emptyResult`, because the Compat collection returns an empty array, not `nil`. I added one more check that if the len compat result is 0, the target result must be the same (so then we check that both of them give empty arrays and not something else).

Reply:
Ah yes that makes sense, thanks for the explanation! :hugs: